Notebook to ascertain whether we can incorporate procedures into the context data for generating text.

In [1]:
import pandas as pd
import numpy as np
import os
import psycopg2
import sqlalchemy
import string

In [2]:
dbschema='mimiciii'
cnx = sqlalchemy.create_engine('postgresql+psycopg2://aa5118:mimic@10.45.20.20:5432/mimic',
                    connect_args={'options': '-csearch_path={}'.format(dbschema)})


In [3]:
test = pd.read_sql_query('''select * from procedures_icd''', cnx)

In [4]:
# procedures data

df_procedures = pd.read_sql_query('''
  SELECT p.subject_id, p.hadm_id, p.seq_num, p.icd9_code, icd.short_title, icd.long_title
  FROM procedures_icd p
  INNER JOIN d_icd_procedures icd 
  USING (icd9_code)
  ORDER BY p.subject_id, p.seq_num
  --LIMIT 10000;
''', cnx)

print(df_procedures.shape)
df_procedures.head(20)

(237948, 6)


subject_id  hadm_id  seq_num icd9_code               short_title  \
0            2   163353        1      9955           Vaccination NEC   
1            3   145834        1      9604  Insert endotracheal tube   
2            3   145834        2      9962    Heart countershock NEC   
3            3   145834        3      8964  Pulmon art wedge monitor   
4            3   145834        4      9672  Cont inv mec ven 96+ hrs   
5            3   145834        5      3893           Venous cath NEC   
6            3   145834        6       966   Entral infus nutrit sub   
7            4   185777        1      3893           Venous cath NEC   
8            4   185777        2      8872       Dx ultrasound-heart   
9            4   185777        3      3323        Other bronchoscopy   
10           5   178980        1      9955           Vaccination NEC   
11           6   107064        1      5569     Kidney transplant NEC   
12           6   107064        2      0091   Trnsplnt live rel donor   
13           6   107064        3      3957    Rep vess w synth patch   
14           6   107064        4      3806   Abdomen artery incision   
15           6   107064        5      9904   Packed cell transfusion   
16           7   118037        1      9955           Vaccination NEC   
17           8   159514        1      9390    Non-invasive mech vent   
18           8   159514        2       640              Circumcision   
19           8   159514        3      9955           Vaccination NEC   

                                           long_title  
0   Prophylactic administration of vaccine against...  
1                      Insertion of endotracheal tube  
2                Other electric countershock of heart  
3                   Pulmonary artery wedge monitoring  
4   Continuous invasive mechanical ventilation for...  
5    Venous catheterization, not elsewhere classified  
6   Enteral infusion of concentrated nutritional s...  
7    Venous catheterization, not elsewhere classified  
8                      Diagnostic ultrasound of heart  
9                                  Other bronchoscopy  
10  Prophylactic administration of vaccine against...  
11                       Other kidney transplantation  
12                 Transplant from live related donor  
13  Repair of blood vessel with synthetic patch graft  
14             Incision of vessel, abdominal arteries  
15                        Transfusion of packed cells  
16  Prophylactic administration of vaccine against...  
17                Non-invasive mechanical ventilation  
18                                       Circumcision  
19  Prophylactic administration of vaccine against...

In [5]:
# procedures data

df_proc_sum = pd.read_sql_query('''
  SELECT subject_id, hadm_id, COUNT(seq_num)
  FROM procedures_icd p
  INNER JOIN d_icd_procedures icd 
  USING (icd9_code)
  GROUP BY subject_id, hadm_id
  ORDER BY COUNT(seq_num) DESC
  --LIMIT 10000;
''', cnx)

print(df_proc_sum.shape)
df_proc_sum.head(20)

(52219, 3)


subject_id  hadm_id  count
0        29467   194819     40
1        57764   103584     40
2        62795   173748     40
3        24810   143994     40
4        27755   155889     38
5        99544   145236     38
6        10187   138921     37
7        17029   190907     36
8         1427   178731     36
9        40548   168670     36
10       48149   102223     36
11       81728   179517     35
12       99213   149621     35
13       59980   108685     34
14       32476   119862     34
15       64334   191708     33
16       26492   125102     33
17       82649   194291     32
18       94889   147635     32
19       20920   148445     32

In [6]:
np.mean(df_proc_sum['count'])

4.556732223903177

Each subject has on average 4.5 procedures during a hospital admission

In [7]:
df_adm = pd.read_sql_query('''
  SELECT n.subject_id, n.hadm_id, MAX(n.chartdate) AS chartdate, MAX(a.dischtime) AS dischtime,
  (cast(MAX(a.dischtime) as date) - cast(MAX(n.chartdate) as date)) AS diff
  FROM noteevents n
  JOIN admissions a
  ON n.subject_id = a.subject_id AND n.hadm_id = a.hadm_id 
  WHERE n.category = 'Discharge summary'
  GROUP BY n.subject_id, n.hadm_id
  ORDER BY diff DESC
  --LIMIT 10000;
''', cnx)

print(df_adm.shape)
df_adm.head(20)

(52726, 5)


subject_id  hadm_id  chartdate           dischtime  diff
0         6145   163196 2110-06-13 2110-09-02 15:35:00    81
1        15482   178068 2125-01-12 2125-02-20 13:45:00    39
2        22560   137561 2164-04-10 2164-05-18 19:00:00    38
3         6764   138049 2159-01-30 2159-03-08 14:00:00    37
4          632   104207 2159-11-01 2159-12-02 14:55:00    31
5        15006   155765 2189-12-23 2190-01-23 12:25:00    31
6        26446   150657 2152-04-07 2152-05-08 12:00:00    31
7         4604   113052 2116-04-15 2116-05-15 12:00:00    30
8         7648   118565 2130-09-27 2130-10-24 11:42:00    27
9        16727   157755 2158-09-01 2158-09-27 16:11:00    26
10       15114   192621 2133-07-15 2133-08-09 15:15:00    25
11       21389   175837 2182-09-06 2182-10-01 10:55:00    25
12       11973   135902 2105-02-19 2105-03-13 12:30:00    22
13       18103   108435 2183-03-18 2183-04-07 15:20:00    20
14        5246   119872 2185-12-02 2185-12-21 18:15:00    19
15        4534   146897 2139-01-12 2139-01-30 19:13:00    18
16       19533   139341 2108-01-20 2108-02-07 14:13:00    18
17       24072   149120 2102-08-22 2102-09-08 10:30:00    17
18       21431   100264 2139-09-06 2139-09-23 14:25:00    17
19        1697   100545 2148-12-13 2148-12-30 18:50:00    17

In [8]:
df_adm.loc[df_adm['diff'] < 0, 'diff'] = 0

In [9]:
np.mean(df_adm['diff'])

0.029435193263285666

Clearly there are a lot of patients who stay for a long time in hospital after being discharged from the ICU. However the average stay post ICU discharge is only 0.03 days. Meaning almost all patients leave hospital after being discharged from the ICU. What about if we discount patients who died in ICU.

In [10]:
df_death = pd.read_sql_query('''
  SELECT subject_id, dod
  FROM patients
  ORDER BY subject_id
  --LIMIT 10000;
''', cnx)

print(df_death.shape)
df_death.head()

(46520, 2)


subject_id        dod
0           2        NaT
1           3 2102-06-14
2           4        NaT
3           5        NaT
4           6        NaT

In [11]:
df_temp = pd.merge(df_adm, df_death,  how='left', left_on=['subject_id'], right_on = ['subject_id'])
df_temp.head()

subject_id  hadm_id  chartdate           dischtime  diff        dod
0        6145   163196 2110-06-13 2110-09-02 15:35:00    81 2110-10-18
1       15482   178068 2125-01-12 2125-02-20 13:45:00    39        NaT
2       22560   137561 2164-04-10 2164-05-18 19:00:00    38 2164-05-18
3        6764   138049 2159-01-30 2159-03-08 14:00:00    37        NaT
4         632   104207 2159-11-01 2159-12-02 14:55:00    31        NaT

In [12]:
df_temp['dead'] = np.where(df_temp['dod'] <= df_temp['chartdate'], 1, 0)

In [13]:
df_temp2 = df_temp[df_temp['dead'] == 0]
df_temp2.shape

(47422, 7)

In [14]:
np.mean(df_temp2['diff'])

0.03272742608915693

It turns out this only barely makes a difference. Still, most patients are discharged immediately after their ICU stay. I have now realised that we are including neonates in this. Let's remove them to look at only adults.

In [15]:
sql = """
  SELECT DISTINCT p.subject_id
  FROM patients p 
  INNER JOIN noteevents n 
  ON p.subject_id = n.subject_id
  WHERE ROUND((cast(chartdate as date) - cast(dob as date)) / 365.242,0) > 14
  AND n.category = 'Discharge summary'
  ORDER BY subject_id
  --LIMIT 100;
"""

df = pd.read_sql_query(sqlalchemy.text(sql), cnx)
adults = list(df['subject_id'])

In [16]:
df_temp3 = df_temp[df_temp['subject_id'].isin(adults)]
print(df_temp3.shape)
df_temp3.head()

(48902, 7)


subject_id  hadm_id  chartdate           dischtime  diff        dod  dead
0        6145   163196 2110-06-13 2110-09-02 15:35:00    81 2110-10-18     0
2       22560   137561 2164-04-10 2164-05-18 19:00:00    38 2164-05-18     0
6       26446   150657 2152-04-07 2152-05-08 12:00:00    31 2152-05-08     0
8        7648   118565 2130-09-27 2130-10-24 11:42:00    27 2130-10-24     0
9       16727   157755 2158-09-01 2158-09-27 16:11:00    26 2165-02-19     0

In [17]:
np.mean(df_temp3['diff'])

0.017115864381824875

As predicted, the average stay post ICU discharge drops even further after removing neonates. Let's take a closer look at some of these patients who are having super long stays in hospital

In [18]:
df_procedures[df_procedures['hadm_id'] == 150657]

subject_id  hadm_id  seq_num icd9_code               short_title  \
133763       26446   150657        1       311    Temporary tracheostomy   
133764       26446   150657        2      3612  Aortocor bypas-2 cor art   
133765       26446   150657        3       341   Incision of mediastinum   
133766       26446   150657        4      3723   Rt/left heart card cath   
133767       26446   150657        5      8856  Coronar arteriogr-2 cath   
133768       26446   150657        6      8853  Lt heart angiocardiogram   
133769       26446   150657        7      3761  Pulsation balloon implan   
133770       26446   150657        9      4621       Temporary ileostomy   
133771       26446   150657       10      4595          Anal anastomosis   
133772       26446   150657       11       415         Total splenectomy   
133773       26446   150657       12      0014   Injection oxazolidinone   
133774       26446   150657       13      4639           Enterostomy NEC   
133775       26446   150657       14      4462       Gastrostomy closure   

                                               long_title  
133763                             Temporary tracheostomy  
133764    (Aorto)coronary bypass of two coronary arteries  
133765                            Incision of mediastinum  
133766  Combined right and left heart cardiac catheter...  
133767         Coronary arteriography using two catheters  
133768         Angiocardiography of left heart structures  
133769                       Implant of pulsation balloon  
133770                                Temporary ileostomy  
133771                                Anastomosis to anus  
133772                                  Total splenectomy  
133773  Injection or infusion of oxazolidinone class o...  
133774                                  Other enterostomy  
133775                             Closure of gastrostomy

There aren't that many procedures, but still to be safe, we can only include procedures for patients who leave hopsital within a day of discharge.

In [19]:
df_final = df_temp3[df_temp3['diff'] <= 1]
print(df_final.shape)
df_final.head()

(48803, 7)


subject_id  hadm_id  chartdate           dischtime  diff        dod  dead
213       43430   131076 2182-07-08 2182-07-09 10:46:00     1        NaT     0
214       16412   171732 2195-09-07 2195-09-08 12:30:00     1 2198-08-26     0
215        4826   101492 2174-06-11 2174-06-12 14:09:00     1 2174-08-02     0
216       22115   102011 2191-03-24 2191-03-25 14:11:00     1        NaT     0
217        3911   155474 2141-03-29 2141-03-30 12:03:00     1 2141-04-17     0

In [20]:
df_final.hadm_id.unique().shape[0]

48803

The total number of unique hospital admission ids in our discharge summary dataset is 48,902. So we are only losing out on 99 discharge summaries for context data - not the end of the world by any means. Let's export the hospital admission ids to use in our preprocessing. However, we'll still only use it only for the last discharge note of an admission - just to be safe

In [21]:
df_final[['hadm_id']].to_csv('../data/df_proc_hadm_ids.csv',index=False)